aten/src/ATen/cuda/CUDAEvent.h

```diff
@@ -72,10 +72,14 @@ struct AT_CUDA_API CUDAEvent {
   }
 
   void record(const CUDAStream& stream) {
+    at::cuda::CUDAGuard device_index_guard(static_cast<int16_t>(stream.device_index()));
+    
     if (is_created_) {
       AT_ASSERT(device_index_ == stream.device_index());
     } else {
-      create(stream.device_index());
+      AT_CUDA_CHECK(cudaEventCreateWithFlags(&event_, flags_));
+      is_created_ = true;
+      device_index_ = stream.device_index();
     }
 
     AT_CUDA_CHECK(cudaEventRecord(event_, stream));
```

Comment:
I suppose this wasn't expected to work in the first place if the current device was not the right one, hence the device guard? cc @ezyang

Reply:
Turns out both cudaEventRecord() and cudaStreamWaitEvent() need guards that set the device to the stream's device.

Reply:
I added some comments for this. The creation can happen on any device. The record has to happen on the stream's device.